src/tokens/ERC1155.sol

```diff
@@ -0,0 +1,192 @@
+  /// @param _amount   Transfered amount
+  /// @param _gasLimit The gas limit
+  /// @param _data     Data to call with
+  function _callonERC1155Received(address _from, address _to, uint256 _id, uint256 _amount, uint256 _gasLimit, bytes memory _data) internal {
+    // Check if recipient is contract using inline code length
+    if (_to.code.length > 0) {
+      bytes4 retval = IERC1155TokenReceiver(_to).onERC1155Received{gas: _gasLimit}(msg.sender, _from, _id, _amount, _data);
+      require(retval == ERC1155_RECEIVED_VALUE, "ERC1155#_callonERC1155Received: INVALID_ON_RECEIVE_MESSAGE");
+    }
+  }
+
+  /// @notice Send multiple types of Tokens from the _from address to the _to address (with safety call)
+  /// @param _from     Source addresses
+  /// @param _to       Target addresses
+  /// @param _ids      IDs of each token type
+  /// @param _amounts  Transfer amounts per token type
+  function _safeBatchTransferFrom(address _from, address _to, uint256[] memory _ids, uint256[] memory _amounts) internal {
+    require(_ids.length == _amounts.length, "ERC1155#_safeBatchTransferFrom: INVALID_ARRAYS_LENGTH");
+
+    // Number of transfer to execute
+    uint256 nTransfer = _ids.length;
+
+    // Executing all transfers
+    for (uint256 i = 0; i < nTransfer; i++) {
+      // Update storage balance of previous bin
+      balances[_from][_ids[i]] = balances[_from][_ids[i]].sub(_amounts[i]);
+      balances[_to][_ids[i]] = balances[_to][_ids[i]].add(_amounts[i]);
+    }
+
+    emit TransferBatch(msg.sender, _from, _to, _ids, _amounts);
+  }
+
+  /// @notice Verifies if receiver is contract and if so, calls (_to).onERC1155BatchReceived(...)
+  /// @param _from     Source addresses
+  /// @param _to       Target addresses
+  /// @param _ids      IDs of each token type
+  /// @param _amounts  Transfer amounts per token type
+  /// @param _gasLimit The gas limit
+  /// @param _data     Data to call with
+  function _callonERC1155BatchReceived(address _from, address _to, uint256[] memory _ids, uint256[] memory _amounts, uint256 _gasLimit, bytes memory _data) internal {
+    // Pass data if recipient is contract
+    if (_to.code.length > 0) {
+      bytes4 retval = IERC1155TokenReceiver(_to).onERC1155BatchReceived{gas: _gasLimit}(msg.sender, _from, _ids, _amounts, _data);
+      require(retval == ERC1155_BATCH_RECEIVED_VALUE, "ERC1155#_callonERC1155BatchReceived: INVALID_ON_RECEIVE_MESSAGE");
+    }
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                        Operator Functions
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice Enable or disable approval for a third party ("operator") to manage all of caller's tokens
+  /// @param _operator  Address to add to the set of authorized operators
+  /// @param _approved  True if the operator is approved, false to revoke approval
+  function setApprovalForAll(address _operator, bool _approved) external override {
```

Comment:
we dont use underlines for variable names :)

Reply:
okie

Reply:
sry shoulda drafted pr

Reply:
will conform stronger rn

Reply:
nw nw idrc if its draft or not lol